File: src/main/java/io/austinzhu/type/functor/data/Sum.java
package io.austinzhu.type.functor.data;

import io.austinzhu.type.Type;
import io.austinzhu.type.functor.Functor;

public interface Sum<
        F extends Functor<F, A> & Type<F>,
        G extends Functor<G, A> & Type<G>,
        A extends Type<A>> {

    class InL<
            F extends Functor<F, A> & Type<F>,
            G extends Functor<G, A> & Type<G>,
            A extends Type<A>> implements Sum<F, G, A> {

        F f;

        InL(F f) {
            this.f = f;
        }
    }

    class InR<
            F extends Functor<F, A> & Type<F>,
            G extends Functor<G, A> & Type<G>,
            A extends Type<A>> implements Sum<F, G, A> {

        G g;

        InR(G g) {
            this.g = g;
        }
    }
}

File: src/main/scala/dev/austinzhu/pattern/functor/data/Sum.scala
package dev.austinzhu.pattern.functor.data

type.Type

object Sum {
  class InL[F <: Functor[F, A] with Type, G <: Functor[G, A] with Type, A <: Type] private[data](var f: F) extends Sum[F, G, A] {
  }

  class InR[F <: Functor[F, A] with Type, G <: Functor[G, A] with Type, A <: Type] private[data](var g: G) extends Sum[F, G, A] {
  }
}